#! /bin/sh /usr/share/dpatch/dpatch-run
## 06_nissl_dvbplayer.dpatch by Reinhard Nissl <rnissl@gmx.de>
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: send proper I-frames to the device

@DPATCH@
diff -urNad vdr-1.3.24/dvbplayer.c /tmp/dpep.Ctc0VI/vdr-1.3.24/dvbplayer.c
--- vdr-1.3.24/dvbplayer.c	2005-05-08 16:52:49.000000000 +0200
+++ /tmp/dpep.Ctc0VI/vdr-1.3.24/dvbplayer.c	2005-05-09 20:31:45.721116992 +0200
@@ -360,6 +360,94 @@
      }
 }
 
+static uchar *findStartCode(uchar *Data, int Length)
+{
+  uchar *limit = Data + Length;
+  Data += 6 + 3 + Data[6 + 2] + 3; // move to video payload and skip 00 00 01
+  while (Data < limit) {
+        // possible start codes that appear before/after picture data
+        // 00 00 01 B3: sequence header code
+        // 00 00 01 B8: group start code
+        // 00 00 01 00: picture start code
+        // 00 00 01 B7: sequence end code
+        if (0x01 == Data[-1] && (0xB3 == Data[0] || 0xB8 == Data[0] || 0x00 == Data[0] || 0xB7 == Data[0]) && 0x00 == Data[-2] && 0x00 == Data[-3])
+            return Data - 3;
+        Data++;
+        }
+
+  return 0;
+}
+
+static void fixIFrameHead(uchar *Data, int Length)
+{
+  uchar *p = findStartCode(Data, Length);
+  if (!p) {
+     esyslog("fixIframeHead: start code not found!\n");
+     return;
+     }
+
+  Data += 6 + 3 + Data[6 + 2]; // move to video payload
+  if (Data < p)
+     memset(Data, 0, p - Data); // zero preceeding bytes
+}
+
+static int fixIFrameTail(uchar *Data, int Length, bool StillImage)
+{
+  uchar *p = findStartCode(Data, Length);
+  if (!p) {
+     esyslog("fixIframeTail: start code not found!\n");
+     return Length;
+     }
+
+  if (StillImage) {
+     // need to append sequence end code
+     p[3] = 0xB7;
+     p += 4;
+     }
+
+  // is this PES packet required?
+  uchar *videoPayload = Data + 6 + 3 + Data[6 + 2];
+  if (videoPayload >= p)
+     return 0; // no
+  
+  // adjust PES payload size
+  int payloadSize = (p - Data) - 6;
+  Data[4] = payloadSize >> 8;
+  Data[5] = payloadSize & 0xFF;
+
+  return payloadSize + 6;
+}
+
+#define IPACKS 2048 // originally defined in remux.c
+
+static void fixIFrame(uchar *Data, int &Length, bool StillImage)
+{
+  int originalLength = Length - IPACKS; // by design
+  int done = 0;
+
+  while (done < Length) {
+        if (0x00 != Data[0] || 0x00 != Data[1] || 0x01 != Data[2]) {
+           esyslog("fixIFrame: PES start code not found at offset %d (data length: %d)!", done, Length);
+           return;
+           }
+
+        int lenPES = 6 + Data[4] * 256 + Data[5];
+        if (0xE0 == (0xF0 & Data[ 3 ])) { // video packet
+           int todo = Length - done;
+           int bite = (lenPES < todo) ? lenPES : todo;
+           if (0 == done) // first packet
+              fixIFrameHead(Data, bite);
+           else if (done >= originalLength) { // last packet
+              Length = done + fixIFrameTail(Data, bite, StillImage);
+              return;
+              }
+           }
+        
+        done += lenPES;
+        Data += lenPES;
+        }
+}
+
 void cDvbPlayer::Action(void)
 {
   uchar *b = NULL;
@@ -395,6 +483,7 @@
                        int FileOffset;
                        int Index = index->GetNextIFrame(readIndex, playDir == pdForward, &FileNumber, &FileOffset, &Length, true);
                        if (Index >= 0) {
+                          Length += IPACKS; // fixIFrame needs next video packet
                           if (!NextFile(FileNumber, FileOffset))
                              continue;
                           }
@@ -435,6 +524,8 @@
                     }
                  int r = nonBlockingFileReader->Read(replayFile, b, Length);
                  if (r > 0) {
+                    if (playMode == pmFast || (playMode == pmSlow && playDir == pdBackward))
+                       fixIFrame(b, r, false);
                     readFrame = new cFrame(b, -r, ftUnknown, readIndex); // hands over b to the ringBuffer
                     b = NULL;
                     }
@@ -666,11 +757,13 @@
      int FileOffset, Length;
      Index = index->GetNextIFrame(Index, false, &FileNumber, &FileOffset, &Length);
      if (Index >= 0 && NextFile(FileNumber, FileOffset) && Still) {
+        Length += IPACKS; // fixIFrame needs next video packet
         uchar b[MAXFRAMESIZE];
         int r = ReadFrame(replayFile, b, Length, sizeof(b));
         if (r > 0) {
            if (playMode == pmPause)
               DevicePlay();
+           fixIFrame(b, r, true);
            DeviceStillPicture(b, r);
            }
         playMode = pmStill;
